drivers/timer/arcv2_timer0.c

```diff
@@ -156,323 +108,44 @@ static ALWAYS_INLINE void timer0_limit_register_set(u32_t count)
 	_arc_v2_aux_reg_write(_ARC_V2_TMR0_LIMIT, count);
 }
 
-#ifdef CONFIG_TICKLESS_IDLE
-static ALWAYS_INLINE void update_accumulated_count(void)
-{
-	accumulated_cycle_count += (_sys_idle_elapsed_ticks * cycles_per_tick);
-}
-#else /* CONFIG_TICKLESS_IDLE */
-static ALWAYS_INLINE void update_accumulated_count(void)
+static u32_t elapsed(void)
 {
-	accumulated_cycle_count += cycles_per_tick;
-}
-#endif /* CONFIG_TICKLESS_IDLE */
+	u32_t val, ov;
 
-#ifdef CONFIG_TICKLESS_KERNEL
-static inline void program_max_cycles(void)
-{
-	timer0_limit_register_set(max_system_ticks * cycles_per_tick);
-	timer_expired = 0;
+	do {
+		val =  timer0_count_register_get();
+		ctrl_cache = timer0_control_register_get();
+	} while (timer0_count_register_get() < val);
+
+	ov = (ctrl_cache & _ARC_V2_TMR_CTRL_IP) ? last_load : 0;
+	return val + ov;
 }
-#endif
 
 /**
  *
  * @brief System clock periodic tick handler
  *
- * This routine handles the system clock periodic tick interrupt. It always
- * announces one tick.
+ * This routine handles the system clock tick interrupt. It always
+ * announces one tick when TICKLESS is not enabled, or multiple ticks
+ * when TICKLESS is enabled.
  *
  * @return N/A
  */
 void _timer_int_handler(void *unused)
```

Comment:
Unless this is required for external linkage, make this a static function.  A few architectures are still weird, but it would be good to get away from "the timer ISR always has a particular name".

Reply:
agreed, it should be static, i will change it.